src/bun.js/api/bun/h2_frame_parser.zig

```diff
@@ -0,0 +1,2440 @@
+        this.onGoAway.protect();
+        this.onFrameError.protect();
+        this.onAborted.protect();
+    }
+};
+
+pub const H2FrameParser = struct {
+    pub const log = Output.scoped(.H2FrameParser, false);
+    pub usingnamespace JSC.Codegen.JSH2FrameParser;
+
+    strong_ctx: JSC.Strong = .{},
+    allocator: Allocator,
+    handlers: Handlers,
+    localSettings: FullSettingsPayload = .{},
+    // only available after receiving settings or ACK
+    remoteSettings: ?FullSettingsPayload = null,
+    // current frame being read
+    currentFrame: ?FrameHeader = null,
+    // remaining bytes to read for the current frame
+    remainingLength: i32 = 0,
+    // buffer if more data is needed for the current frame
+    readBuffer: MutableString,
+    // current window size for the connection
+    windowSize: u32 = 65535,
+    // used window size for the connection
+    usedWindowSize: u32 = 0,
+    lastStreamID: u32 = 0,
+    streams: bun.U32HashMap(Stream),
+
+    decoder: lshpack.lshpack_dec = undefined,
+    encoder: lshpack.lshpack_enc = undefined,
+
+    const Stream = struct {
+        id: u32 = 0,
+        state: enum(u8) {
+            IDLE = 1,
+            RESERVED_LOCAL = 3,
+            RESERVED_REMOTE = 4,
+            OPEN = 2,
+            HALF_CLOSED_LOCAL = 5,
+            HALF_CLOSED_REMOTE = 6,
+            CLOSED = 7,
+        } = .IDLE,
+        waitForTrailers: bool = false,
+        endAfterHeaders: bool = false,
+        isWaitingMoreHeaders: bool = false,
+        padding: ?u8 = 0,
+        rstCode: u32 = 0,
+        streamDependency: u32 = 0,
+        exclusive: bool = false,
+        weight: u16 = 36,
+        // current window size for the stream
+        windowSize: u32 = 65535,
+        // used window size for the stream
+        usedWindowSize: u32 = 0,
+
+        signal: ?*JSC.WebCore.AbortSignal = null,
+        client: *H2FrameParser,
+
+        pub fn init(streamIdentifier: u32, initialWindowSize: u32, client: *H2FrameParser) Stream {
+            var stream = Stream{
+                .id = streamIdentifier,
+                .state = .OPEN,
+                .windowSize = initialWindowSize,
+                .usedWindowSize = 0,
+                .weight = 36,
+                .client = client,
+            };
+            return stream;
+        }
+
+        pub fn canReceiveData(this: *Stream) bool {
+            return switch (this.state) {
+                .IDLE, .RESERVED_LOCAL, .RESERVED_REMOTE, .OPEN, .HALF_CLOSED_LOCAL => false,
+                .HALF_CLOSED_REMOTE, .CLOSED => true,
+            };
+        }
+
+        pub fn canSendData(this: *Stream) bool {
+            return switch (this.state) {
+                .IDLE, .RESERVED_LOCAL, .RESERVED_REMOTE, .OPEN, .HALF_CLOSED_REMOTE => false,
+                .HALF_CLOSED_LOCAL, .CLOSED => true,
+            };
+        }
+
+        pub fn attachSignal(this: *Stream, signal: *JSC.WebCore.AbortSignal) void {
+            _ = signal.ref();
```

Comment:
maybe we should make the return type of `ref` void because we never use it

Reply:
that or we make this 

```suggestion
            this.signal = signal.ref().listen(Stream, this, Stream.abortListener);
```